lib/geordi/db_cleaner.rb

```diff
@@ -0,0 +1,183 @@
+      return if deletable_dbs.nil?
+      deletable_dbs.each do |db|
+        note "Dropping database `#{db}`."
+        `#{postgres_command} -c 'DROP DATABASE "#{db}";'`
+      end
+    end
+
+    def list_all_mysql_dbs(mysql_command)
+      `#{mysql_command} -B -N -e 'show databases'`.split
+    end
+    private :list_all_mysql_dbs
+
+    def whitelist_fname(dbtype)
+      File.join(@whitelist_directory, dbtype) << '.yaml'
+    end
+
+    def confirm_deletion(dbtype, database_list)
+      proceed = ''
+      until %w[y n].include? proceed
+        deletable_dbs = filter_whitelisted(dbtype, database_list)
+        if deletable_dbs.empty?
+          note "No #{dbtype} databases found that were not whitelisted"
+          if prompt('Edit the whitelist? [y]es or [n]o') == 'y'
+            proceed = 'e'
+          else
+            return []
+          end
+        end
+        if proceed.empty?
+          note "The following #{dbtype} databases are not whitelisted and will be deleted:"
+          deletable_dbs.sort.each do |db|
+            note db
+          end
+          proceed = prompt('Proceed? [y]es, [n]o or [e]dit whitelist')
+        end
+        case proceed
+        when 'e'
+          proceed = ''  # reset user selection
+          edit_whitelist dbtype
+        when 'n'
+          announce 'Not deleting anything'
+          return []
+        when 'y'
+          return deletable_dbs
+        end
+      end
+    end
+    private :confirm_deletion
+
+    def create_whitelist(dbtype)
+      whitelist = File.open(whitelist_fname(dbtype), 'w')
+      if dbtype == 'mysql'
+        whitelist.write("# Always whitelisted:\n# information_schema\n# performance_schema\n# mysql\n# sys\n")
+      elsif dbtype == 'postgres'
+        whitelist.write("# Always whitelisted: \n # postgres\n")
+      end
+      whitelist.write("# When you whitelist `foo`, `foo_development` and `foo_test\d?` will be considered whitelisted, too.")
+      whitelist.close
+    end
+
+    def filter_whitelisted(dbtype, database_list)
+      create_whitelist(dbtype) unless File.exist? whitelist_fname(dbtype)
+      protected = {
+        'mysql'    => %w[mysql information_schema performance_schema sys],
+        'postgres' => ['postgres'],
+      }
+      whitelist_content = File.open(whitelist_fname(dbtype), 'r').read.lines.map(&:chomp).map(&:strip)
+      # n.b. `delete` means 'delete from list of dbs that should be deleted in this context
+      # i.e. `delete` means 'keep this database'
+      deletable_dbs = database_list.dup
+      deletable_dbs.delete_if { |db| whitelist_content.include? db.sub(/_(test\d?|development)$/, '') }
+      deletable_dbs.delete_if { |db| protected[dbtype].include? db }
+      deletable_dbs.delete_if { |db| db.start_with? '#' }
+    end
+    private :filter_whitelisted
+  end
+end
+
+def choose_texteditor
+  %w[$VISUAL $EDITOR /usr/bin/editor vi].each do |texteditor|
+    return texteditor if cmd_exists? texteditor
+  end
+end
```

Comment:
These methods are globally defined. Please prefer to either include them into the database cleaner class, or move them to Geordi::Util.